engine/access/rpc/backend/backend_network.go

```diff
@@ -23,108 +20,100 @@ type backendNetwork struct {
+
+		// The block exists, but no block has been finalized at its height. Therefore, this block
+		// may be finalized in the future, and the client can retry.
+		return nil, access.NewPreconditionFailedError(
+			fmt.Errorf("failed to retrieve snapshot for block with height %d: block not finalized and is above finalized height",
+				snapshotHeadByBlockId.Height))
 	}
 
 	if blockIDFinalizedAtHeight != blockID {
 		// A different block than what was queried has been finalized at this height.
 		// Therefore, the queried block will never be finalized.
-		return nil, status.Errorf(codes.InvalidArgument,
-			"failed to retrieve snapshot for block: block not finalized and is below finalized height")
+		return nil, access.NewInvalidRequestError(fmt.Errorf("failed to retrieve snapshot for block: block not finalized and is below finalized height"))
 	}
 
 	data, err := convert.SnapshotToBytes(snapshot)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to convert snapshot to bytes: %v", err)
+		return nil, access.RequireNoError(ctx, fmt.Errorf("failed to convert snapshot to bytes: %w", err))
 	}
 	return data, nil
 }
 
 // GetProtocolStateSnapshotByHeight returns serializable Snapshot by block height.
 // The block must be finalized (otherwise the by-height query is ambiguous).
+//
 // Expected errors during normal operation:
-//   - status.Error[codes.NotFound] - No block with the given height was found.
-//     The block height may or may not be finalized in the future; the client can retry later.
-func (b *backendNetwork) GetProtocolStateSnapshotByHeight(_ context.Context, blockHeight uint64) ([]byte, error) {
+//   - access.DataNotFoundError - No finalized block with the given height was found.
+//
+// All errors can be considered benign. Exceptions are handled explicitly within the backend and are
+// not propagated.
+//
+// The block height may or may not be finalized in the future; the client can retry later.
+func (b *backendNetwork) GetProtocolStateSnapshotByHeight(ctx context.Context, blockHeight uint64) ([]byte, error) {
 	snapshot := b.state.AtHeight(blockHeight)
 	_, err := snapshot.Head()
 	if err != nil {
-		if errors.Is(err, state.ErrUnknownSnapshotReference) {
-			return nil, status.Errorf(codes.NotFound, "failed to find snapshot: %v", err)
-		}
-		return nil, status.Errorf(codes.Internal, "failed to get a valid snapshot: %v", err)
+		// storage.ErrNotFound is specifically NOT allowed since the snapshot's reference block must exist
+		// within the snapshot.
+		err = access.RequireErrorIs(ctx, err, state.ErrUnknownSnapshotReference)
+		return nil, access.NewDataNotFoundError("snapshot", fmt.Errorf("failed to retrieve a valid snapshot: block not found"))
 	}
 
 	data, err := convert.SnapshotToBytes(snapshot)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to convert snapshot to bytes: %v", err)
+		return nil, access.RequireNoError(ctx, fmt.Errorf("failed to convert snapshot to bytes: %w", err))
```

Comment:
Only if we can guarantee that we are _always_ be able to generate a snapshot for any block (including the root block when the node is freshly bootstrapped) we can use ` access.RequireNoError` here (same as [comment above](https://github.com/onflow/flow-go/pull/7106#discussion_r1996577791))

```suggestion
		err = access.RequireErrorIs(ctx, err, protocol.ErrSealingSegmentBelowRootBlock, protocol.NewUnfinalizedSealingSegmentErrorf(""))
		return nil, fmt.Errorf("snapshots might not be possible for every block: %w", err)
```